File: backend/app/api/schemas/artifacts.py
from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Any, Dict, Literal, Optional

from pydantic import AliasChoices, BaseModel, ConfigDict, Field, model_validator

class ArtifactType(str, Enum):
    DRAFT = "draft"
    PUBLISHED = "published"


class ArtifactKind(str, Enum):
    AGENT_NODE = "agent_node"
    RAG_OPERATOR = "rag_operator"
    TOOL_IMPL = "tool_impl"


class ArtifactLanguage(str, Enum):
    PYTHON = "python"
    JAVASCRIPT = "javascript"


class ArtifactOwnerType(str, Enum):
    TENANT = "tenant"
    SYSTEM = "system"


class ArtifactSourceFile(BaseModel):
    model_config = ConfigDict(extra="forbid")
    path: str
    content: str


class ArtifactRuntimeConfig(BaseModel):
    model_config = ConfigDict(extra="forbid")
    language: ArtifactLanguage = ArtifactLanguage.PYTHON
    source_files: list[ArtifactSourceFile]
    entry_module_path: str
    dependencies: list[str] = Field(
        default_factory=list,
        validation_alias=AliasChoices("dependencies", "python_dependencies"),
    )
    runtime_target: str = "cloudflare_workers"


class ArtifactCapabilityConfig(BaseModel):
    model_config = ConfigDict(extra="forbid")
    network_access: bool = False
    allowed_hosts: list[str] = Field(default_factory=list)
    secret_refs: list[str] = Field(default_factory=list)
    storage_access: list[str] = Field(default_factory=list)
    side_effects: list[str] = Field(default_factory=list)


class AgentArtifactContract(BaseModel):
    model_config = ConfigDict(extra="forbid")
    state_reads: list[str] = Field(default_factory=list)
    state_writes: list[str] = Field(default_factory=list)
    input_schema: Dict[str, Any] = Field(default_factory=dict)
    output_schema: Dict[str, Any] = Field(default_factory=dict)
    node_ui: Dict[str, Any] = Field(default_factory=dict)


class RAGArtifactContract(BaseModel):
    model_config = ConfigDict(extra="forbid")
    operator_category: str
    pipeline_role: str
    input_schema: Dict[str, Any] = Field(default_factory=dict)
    output_schema: Dict[str, Any] = Field(default_factory=dict)
    execution_mode: str = "background"


class ToolArtifactContract(BaseModel):
    model_config = ConfigDict(extra="forbid")
    input_schema: Dict[str, Any]
    output_schema: Dict[str, Any]
    side_effects: list[str] = Field(default_factory=list)
    execution_mode: str = "interactive"
    tool_ui: Dict[str, Any] = Field(default_factory=dict)


class ArtifactContractEnvelope(BaseModel):
    model_config = ConfigDict(extra="forbid")
    kind: ArtifactKind
    agent_contract: Optional[AgentArtifactContract] = None
    rag_contract: Optional[RAGArtifactContract] = None
    tool_contract: Optional[ToolArtifactContract] = None

    @model_validator(mode="after")
    def validate_contract_shape(self) -> "ArtifactContractEnvelope":
        if self.kind == ArtifactKind.AGENT_NODE:
            if self.agent_contract is None or self.rag_contract is not None or self.tool_contract is not None:
                raise ValueError("agent_node artifacts require only agent_contract")
        elif self.kind == ArtifactKind.RAG_OPERATOR:
            if self.rag_contract is None or self.agent_contract is not None or self.tool_contract is not None:
                raise ValueError("rag_operator artifacts require only rag_contract")
        elif self.kind == ArtifactKind.TOOL_IMPL:
            if self.tool_contract is None or self.agent_contract is not None or self.rag_contract is not None:
                raise ValueError("tool_impl artifacts require only tool_contract")
        return self


class ArtifactSchema(BaseModel):
    model_config = ConfigDict(extra="forbid")
    id: str
    display_name: str
    description: Optional[str] = None
    kind: ArtifactKind
    owner_type: ArtifactOwnerType
    type: ArtifactType
    version: str
    config_schema: Dict[str, Any] = Field(default_factory=dict)
    runtime: ArtifactRuntimeConfig
    capabilities: ArtifactCapabilityConfig = Field(default_factory=ArtifactCapabilityConfig)
    agent_contract: Optional[AgentArtifactContract] = None
    rag_contract: Optional[RAGArtifactContract] = None
    tool_contract: Optional[ToolArtifactContract] = None
    created_at: Optional[datetime] = None
    updated_at: datetime
    system_key: Optional[str] = None
    author: Optional[str] = None
    tags: list[str] = Field(default_factory=list)


class ArtifactVersionListItem(BaseModel):
    model_config = ConfigDict(extra="forbid")
    id: str
    artifact_id: str
    revision_number: int
    version_label: str
    is_published: bool = False
    is_current_draft: bool = False
    is_current_published: bool = False
    source_file_count: int = 0
    created_at: datetime
    created_by: Optional[str] = None


class ArtifactVersionSchema(ArtifactVersionListItem):
    display_name: str
    description: Optional[str] = None
    kind: ArtifactKind
    config_schema: Dict[str, Any] = Field(default_factory=dict)
    runtime: ArtifactRuntimeConfig
    capabilities: ArtifactCapabilityConfig = Field(default_factory=ArtifactCapabilityConfig)
    agent_contract: Optional[AgentArtifactContract] = None
    rag_contract: Optional[RAGArtifactContract] = None
    tool_contract: Optional[ToolArtifactContract] = None


class ArtifactWorkingDraftResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")
    artifact_id: Optional[str] = None
    draft_key: Optional[str] = None
    draft_snapshot: Dict[str, Any] = Field(default_factory=dict)
    updated_at: Optional[datetime] = None


class ArtifactWorkingDraftUpdateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")
    artifact_id: Optional[str] = None
    draft_key: Optional[str] = None
    draft_snapshot: Dict[str, Any] = Field(default_factory=dict)


class ArtifactCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")
    display_name: str
    description: Optional[str] = None
    draft_key: Optional[str] = None
    kind: ArtifactKind
    runtime: ArtifactRuntimeConfig
    capabilities: ArtifactCapabilityConfig = Field(default_factory=ArtifactCapabilityConfig)
    config_schema: Dict[str, Any] = Field(default_factory=dict)
    agent_contract: Optional[AgentArtifactContract] = None
    rag_contract: Optional[RAGArtifactContract] = None
    tool_contract: Optional[ToolArtifactContract] = None

    @model_validator(mode="after")
    def validate_contracts(self) -> "ArtifactCreate":
        ArtifactContractEnvelope(
            kind=self.kind,
            agent_contract=self.agent_contract,
            rag_contract=self.rag_contract,
            tool_contract=self.tool_contract,
        )
        return self


class ArtifactUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")
    display_name: Optional[str] = None
    description: Optional[str] = None
    draft_key: Optional[str] = None
    runtime: Optional[ArtifactRuntimeConfig] = None
    capabilities: Optional[ArtifactCapabilityConfig] = None
    config_schema: Optional[Dict[str, Any]] = None
    agent_contract: Optional[AgentArtifactContract] = None
    rag_contract: Optional[RAGArtifactContract] = None
    tool_contract: Optional[ToolArtifactContract] = None


class ArtifactConvertKindRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")
    kind: ArtifactKind
    agent_contract: Optional[AgentArtifactContract] = None
    rag_contract: Optional[RAGArtifactContract] = None
    tool_contract: Optional[ToolArtifactContract] = None

    @model_validator(mode="after")
    def validate_contracts(self) -> "ArtifactConvertKindRequest":
        ArtifactContractEnvelope(
            kind=self.kind,
            agent_contract=self.agent_contract,
            rag_contract=self.rag_contract,
            tool_contract=self.tool_contract,
        )
        return self


class ArtifactTestRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")
    artifact_id: Optional[str] = None
    source_files: list[ArtifactSourceFile] = Field(default_factory=list)
    entry_module_path: Optional[str] = None
    input_data: Any
    config: Dict[str, Any] = Field(default_factory=dict)
    dependencies: list[str] = Field(default_factory=list)
    language: Optional[ArtifactLanguage] = None
    kind: Optional[ArtifactKind] = None
    runtime_target: Optional[str] = None
    capabilities: Dict[str, Any] = Field(default_factory=dict)
    config_schema: Dict[str, Any] = Field(default_factory=dict)
    agent_contract: Optional[AgentArtifactContract] = None
    rag_contract: Optional[RAGArtifactContract] = None
    tool_contract: Optional[ToolArtifactContract] = None


class ArtifactSourceValidationRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")
    language: ArtifactLanguage
    source_files: list[ArtifactSourceFile] = Field(default_factory=list)
    dependencies: list[str] = Field(default_factory=list)


class ArtifactSourceValidationDiagnostic(BaseModel):
    model_config = ConfigDict(extra="forbid")
    path: str
    message: str
    line: int = 1
    column: int = 1
    end_line: int = 1
    end_column: int = 1
    severity: Literal["error"] = "error"
    code: Optional[str] = None


class ArtifactSourceValidationResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")
    diagnostics: list[ArtifactSourceValidationDiagnostic] = Field(default_factory=list)


class ArtifactDependencyAnalysisRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")
    language: ArtifactLanguage
    source_files: list[ArtifactSourceFile] = Field(default_factory=list)
    dependencies: list[str] = Field(default_factory=list)


class ArtifactDependencyRow(BaseModel):
    model_config = ConfigDict(extra="forbid")
    name: str
    normalized_name: str
    declared_spec: Optional[str] = None
    classification: Literal["builtin", "runtime_provided", "declared"]
    source: Literal["builtin", "runtime_registry", "runtime_catalog", "declared"]
    status: str
    note: Optional[str] = None
    imported: bool = False
    declared: bool = False
    can_remove: bool = False
    can_add: bool = False
    needs_declaration: bool = False


class ArtifactDependencyAnalysisResponse(BaseModel):
    rows: list[ArtifactDependencyRow] = Field(default_factory=list)


class PythonPackageVerificationRequest(BaseModel):
    package_name: str


class PythonPackageVerificationResponse(BaseModel):
    package_name: str
    normalized_name: str
    status: Literal["exists", "not_found", "invalid", "lookup_failed"]
    exists: bool = False
    error_message: Optional[str] = None


class ArtifactTestResponse(BaseModel):
    success: bool
    data: Optional[Any] = None
    error_message: Optional[str] = None
    execution_time_ms: float = 0.0
    run_id: Optional[str] = None
    error_payload: Optional[Dict[str, Any]] = None
    stdout_excerpt: Optional[str] = None
    stderr_excerpt: Optional[str] = None


class ArtifactPublishResponse(BaseModel):
    artifact_id: str
    revision_id: str
    version: str
    status: Literal["published"] = "published"


class ArtifactTransferPayload(BaseModel):
    model_config = ConfigDict(extra="forbid")
    display_name: str
    description: Optional[str] = None
    kind: ArtifactKind
    runtime: ArtifactRuntimeConfig
    capabilities: ArtifactCapabilityConfig = Field(default_factory=ArtifactCapabilityConfig)
    config_schema: Dict[str, Any] = Field(default_factory=dict)
    agent_contract: Optional[AgentArtifactContract] = None
    rag_contract: Optional[RAGArtifactContract] = None
    tool_contract: Optional[ToolArtifactContract] = None
    published: bool = False

    @model_validator(mode="after")
    def validate_contracts(self) -> "ArtifactTransferPayload":
        ArtifactContractEnvelope(
            kind=self.kind,
            agent_contract=self.agent_contract,
            rag_contract=self.rag_contract,
            tool_contract=self.tool_contract,
        )
        return self


class ArtifactTransferFile(BaseModel):
    model_config = ConfigDict(extra="forbid")
    format: Literal["talmudpedia.artifact"] = "talmudpedia.artifact"
    format_version: Literal[1] = 1
    exported_at: datetime
    artifact: ArtifactTransferPayload


class ArtifactImportResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")
    artifact: ArtifactSchema
    source_published: bool = False


class ArtifactRunStatus(str, Enum):
    QUEUED = "queued"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCEL_REQUESTED = "cancel_requested"
    CANCELLED = "cancelled"


class ArtifactRunSchema(BaseModel):
    id: str
    artifact_id: Optional[str] = None
    revision_id: str
    domain: str
    status: ArtifactRunStatus
    queue_class: str
    result_payload: Optional[Dict[str, Any]] = None
    error_payload: Optional[Dict[str, Any]] = None
    stdout_excerpt: Optional[str] = None
    stderr_excerpt: Optional[str] = None
    duration_ms: Optional[int] = None
    runtime_metadata: Dict[str, Any] = Field(default_factory=dict)
    created_at: datetime
    started_at: Optional[datetime] = None
    finished_at: Optional[datetime] = None


class ArtifactRunCreateResponse(BaseModel):
    run_id: str
    status: ArtifactRunStatus


class ArtifactRuntimeQueueStatusSchema(BaseModel):
    queue_class: str
    active_count: int
    concurrency_limit: int


class ArtifactRunEventSchema(BaseModel):
    id: str
    sequence: int
    timestamp: Optional[datetime] = None
    event_type: str
    payload: Dict[str, Any] = Field(default_factory=dict)

File: frontend-reshet/src/services/artifacts.ts
import { httpClient } from "./http";
import { useAuthStore } from "@/lib/store/useAuthStore";

export type ArtifactType = "draft" | "published";
export type ArtifactKind = "agent_node" | "rag_operator" | "tool_impl";
export type ArtifactOwnerType = "tenant" | "system";
export type ArtifactRunStatus = "queued" | "running" | "completed" | "failed" | "cancel_requested" | "cancelled";

export interface ArtifactSourceFile {
  path: string;
  content: string;
}

export interface ArtifactRuntimeConfig {
  source_files: ArtifactSourceFile[];
  entry_module_path: string;
  python_dependencies: string[];
  runtime_target: string;
}

export interface ArtifactCapabilityConfig {
  network_access: boolean;
  allowed_hosts: string[];
  secret_refs: string[];
  storage_access: string[];
  side_effects: string[];
}

export interface AgentArtifactContract {
  state_reads: string[];
  state_writes: string[];
  input_schema: Record<string, unknown>;
  output_schema: Record<string, unknown>;
  node_ui: Record<string, unknown>;
}

export interface RAGArtifactContract {
  operator_category: string;
  pipeline_role: string;
  input_schema: Record<string, unknown>;
  output_schema: Record<string, unknown>;
  execution_mode: string;
}

export interface ToolArtifactContract {
  input_schema: Record<string, unknown>;
  output_schema: Record<string, unknown>;
  side_effects: string[];
  execution_mode: string;
  tool_ui: Record<string, unknown>;
}

export interface Artifact {
  id: string;
  slug: string;
  display_name: string;
  description?: string;
  kind: ArtifactKind;
  owner_type: ArtifactOwnerType;
  type: ArtifactType;
  version: string;
  config_schema: Record<string, unknown>;
  runtime: ArtifactRuntimeConfig;
  capabilities: ArtifactCapabilityConfig;
  agent_contract?: AgentArtifactContract | null;
  rag_contract?: RAGArtifactContract | null;
  tool_contract?: ToolArtifactContract | null;
  created_at?: string;
  updated_at: string;
  system_key?: string | null;
  author?: string | null;
  tags: string[];
}

export interface ArtifactCreateRequest {
  slug: string;
  display_name: string;
  description?: string;
  kind: ArtifactKind;
  runtime: ArtifactRuntimeConfig;
  capabilities: ArtifactCapabilityConfig;
  config_schema: Record<string, unknown>;
  agent_contract?: AgentArtifactContract;
  rag_contract?: RAGArtifactContract;
  tool_contract?: ToolArtifactContract;
}

export interface ArtifactUpdateRequest {
  display_name?: string;
  description?: string;
  runtime?: ArtifactRuntimeConfig;
  capabilities?: ArtifactCapabilityConfig;
  config_schema?: Record<string, unknown>;
  agent_contract?: AgentArtifactContract;
  rag_contract?: RAGArtifactContract;
  tool_contract?: ToolArtifactContract;
}

export interface ArtifactConvertKindRequest {
  kind: ArtifactKind;
  agent_contract?: AgentArtifactContract;
  rag_contract?: RAGArtifactContract;
  tool_contract?: ToolArtifactContract;
}

export interface ArtifactPublishResponse {
  artifact_id: string;
  revision_id: string;
  version: string;
  status: "published";
}

export interface ArtifactTestRequest {
  artifact_id?: string;
  source_files?: ArtifactSourceFile[];
  entry_module_path?: string;
  dependencies?: string[];
  input_data: unknown;
  config?: Record<string, unknown>;
  kind?: ArtifactKind;
  runtime_target?: string;
  capabilities?: ArtifactCapabilityConfig;
  config_schema?: Record<string, unknown>;
  agent_contract?: AgentArtifactContract;
  rag_contract?: RAGArtifactContract;
  tool_contract?: ToolArtifactContract;
}

export interface ArtifactTestResponse {
  success: boolean;
  data?: unknown;
  error_message?: string;
  execution_time_ms: number;
  run_id?: string;
  error_payload?: Record<string, unknown>;
  stdout_excerpt?: string;
  stderr_excerpt?: string;
}

export interface ArtifactRun {
  id: string;
  artifact_id?: string;
  revision_id: string;
  domain: string;
  status: ArtifactRunStatus;
  queue_class: string;
  result_payload?: Record<string, unknown>;
  error_payload?: Record<string, unknown>;
  stdout_excerpt?: string;
  stderr_excerpt?: string;
  duration_ms?: number;
  runtime_metadata?: Record<string, unknown>;
  created_at: string;
  started_at?: string;
  finished_at?: string;
}

export interface ArtifactVersionListItem {
  id: string;
  artifact_id: string;
  revision_number: number;
  version_label: string;
  is_published: boolean;
  is_current_draft: boolean;
  is_current_published: boolean;
  source_file_count: number;
  created_at: string;
  created_by?: string | null;
}

export interface ArtifactVersion extends ArtifactVersionListItem {
  slug: string;
  display_name: string;
  description?: string | null;
  kind: ArtifactKind;
  config_schema: Record<string, unknown>;
  runtime: ArtifactRuntimeConfig;
  capabilities: ArtifactCapabilityConfig;
  agent_contract?: AgentArtifactContract | null;
  rag_contract?: RAGArtifactContract | null;
  tool_contract?: ToolArtifactContract | null;
}

export interface ArtifactWorkingDraft {
  artifact_id?: string | null;
  draft_key?: string | null;
  draft_snapshot: Record<string, unknown>;
  updated_at?: string | null;
}

export interface ArtifactWorkingDraftUpdateRequest {
  artifact_id?: string;
  draft_key?: string;
  draft_snapshot: Record<string, unknown>;
}

export interface ArtifactRunEvent {
  id: string;
  sequence: number;
  timestamp?: string;
  event_type: string;
  payload: Record<string, unknown>;
}

export interface ArtifactRunCreateResponse {
  run_id: string;
  status: ArtifactRunStatus;
}

export interface ArtifactRunEventsResponse {
  run_id: string;
  event_count: number;
  events: ArtifactRunEvent[];
}

export interface ArtifactCodingStreamEvent {
  version?: string;
  event: string;
  run_id: string;
  seq: number;
  ts: string;
  stage: string;
  payload?: Record<string, unknown>;
  diagnostics?: Array<Record<string, unknown>>;
}

export interface ArtifactCodingRun {
  run_id: string;
  status: string;
  chat_session_id?: string | null;
  artifact_id?: string | null;
  draft_key?: string | null;
  surface?: string | null;
  requested_model_id?: string | null;
  resolved_model_id?: string | null;
  error?: string | null;
  created_at: string;
  started_at?: string | null;
  completed_at?: string | null;
}

export type ArtifactCodingScopeMode = "locked" | "standalone";

export interface ArtifactCodingChatSession {
  id: string;
  title: string;
  artifact_id?: string | null;
  draft_key?: string | null;
  scope_mode: ArtifactCodingScopeMode;
  active_run_id?: string | null;
  last_run_id?: string | null;
  created_at: string;
  updated_at: string;
  last_message_at: string;
}

export interface ArtifactCodingChatMessage {
  id: string;
  run_id: string;
  role: "user" | "assistant" | string;
  content: string;
  created_at: string;
}

export interface ArtifactCodingRunEvent {
  run_id: string;
  event: string;
  stage: string;
  payload: Record<string, unknown>;
  diagnostics: Array<Record<string, unknown>>;
  ts?: string | null;
}

export interface ArtifactCodingChatSessionDetail {
  session: ArtifactCodingChatSession;
  messages: ArtifactCodingChatMessage[];
  run_events: ArtifactCodingRunEvent[];
  draft_snapshot: Record<string, unknown>;
  paging: {
    has_more: boolean;
    next_before_message_id?: string | null;
  };
}

export interface ArtifactCodingActiveRunState {
  run_id: string;
  status: string;
}

export interface ArtifactCodingAnswerQuestionRequest {
  question_id: string;
  answers: string[][];
}

export interface ArtifactCodingRevertRequest {
  run_id: string;
}

export interface ArtifactCodingPromptRequest {
  input: string;
  chat_session_id?: string;
  artifact_id?: string;
  draft_key?: string;
  scope_mode?: ArtifactCodingScopeMode;
  model_id?: string | null;
  client_message_id?: string;
  draft_snapshot: Record<string, unknown>;
}

export interface ArtifactCodingPromptSubmissionResponse {
  submission_status: "started";
  chat_session_id: string;
  run: ArtifactCodingRun;
}

export interface ArtifactCodingModelOption {
  id: string | null;
  label: string;
  description?: string;
}

export const artifactsService = {
  list: async (tenantSlug?: string): Promise<Artifact[]> => {
    const url = tenantSlug ? `/admin/artifacts?tenant_slug=${tenantSlug}` : "/admin/artifacts";
    return httpClient.get<Artifact[]>(url);
  },

  get: async (id: string, tenantSlug?: string): Promise<Artifact> => {
    const url = tenantSlug ? `/admin/artifacts/${id}?tenant_slug=${tenantSlug}` : `/admin/artifacts/${id}`;
    return httpClient.get<Artifact>(url);
  },

  listVersions: async (artifactId: string, tenantSlug?: string): Promise<ArtifactVersionListItem[]> => {
    const url = tenantSlug
      ? `/admin/artifacts/${artifactId}/versions?tenant_slug=${tenantSlug}`
      : `/admin/artifacts/${artifactId}/versions`;
    return httpClient.get<ArtifactVersionListItem[]>(url);
  },

  getVersion: async (artifactId: string, revisionId: string, tenantSlug?: string): Promise<ArtifactVersion> => {
    const url = tenantSlug
      ? `/admin/artifacts/${artifactId}/versions/${revisionId}?tenant_slug=${tenantSlug}`
      : `/admin/artifacts/${artifactId}/versions/${revisionId}`;
    return httpClient.get<ArtifactVersion>(url);
  },

  getWorkingDraft: async (artifactId: string, tenantSlug?: string): Promise<ArtifactWorkingDraft> => {
    const url = tenantSlug
      ? `/admin/artifacts/${artifactId}/working-draft?tenant_slug=${tenantSlug}`
      : `/admin/artifacts/${artifactId}/working-draft`;
    return httpClient.get<ArtifactWorkingDraft>(url);
  },

  updateWorkingDraft: async (
    artifactId: string,
    payload: ArtifactWorkingDraftUpdateRequest,
    tenantSlug?: string,
  ): Promise<ArtifactWorkingDraft> => {
    const url = tenantSlug
      ? `/admin/artifacts/${artifactId}/working-draft?tenant_slug=${tenantSlug}`
      : `/admin/artifacts/${artifactId}/working-draft`;
    return httpClient.put<ArtifactWorkingDraft>(url, payload);
  },

  create: async (data: ArtifactCreateRequest, tenantSlug?: string): Promise<Artifact> => {
    const url = tenantSlug ? `/admin/artifacts?tenant_slug=${tenantSlug}` : "/admin/artifacts";
    return httpClient.post<Artifact>(url, data);
  },

  update: async (id: string, data: ArtifactUpdateRequest, tenantSlug?: string): Promise<Artifact> => {
    const url = tenantSlug ? `/admin/artifacts/${id}?tenant_slug=${tenantSlug}` : `/admin/artifacts/${id}`;
    return httpClient.put<Artifact>(url, data);
  },

  publish: async (id: string, tenantSlug?: string): Promise<ArtifactPublishResponse> => {
    const url = tenantSlug ? `/admin/artifacts/${id}/publish?tenant_slug=${tenantSlug}` : `/admin/artifacts/${id}/publish`;
    return httpClient.post<ArtifactPublishResponse>(url, {});
  },

  convertKind: async (id: string, data: ArtifactConvertKindRequest, tenantSlug?: string): Promise<Artifact> => {
    const url = tenantSlug ? `/admin/artifacts/${id}/convert-kind?tenant_slug=${tenantSlug}` : `/admin/artifacts/${id}/convert-kind`;
    return httpClient.post<Artifact>(url, data);
  },

  delete: async (id: string, tenantSlug?: string): Promise<void> => {
    const url = tenantSlug ? `/admin/artifacts/${id}?tenant_slug=${tenantSlug}` : `/admin/artifacts/${id}`;
    await httpClient.delete(url);
  },

  createTestRun: async (data: ArtifactTestRequest, tenantSlug?: string): Promise<ArtifactRunCreateResponse> => {
    const url = tenantSlug ? `/admin/artifacts/test-runs?tenant_slug=${tenantSlug}` : "/admin/artifacts/test-runs";
    return httpClient.post<ArtifactRunCreateResponse>(url, data);
  },

  getRun: async (runId: string, tenantSlug?: string): Promise<ArtifactRun> => {
    const url = tenantSlug ? `/admin/artifact-runs/${runId}?tenant_slug=${tenantSlug}` : `/admin/artifact-runs/${runId}`;
    return httpClient.get<ArtifactRun>(url);
  },

  getRunEvents: async (runId: string, tenantSlug?: string): Promise<ArtifactRunEventsResponse> => {
    const url = tenantSlug
      ? `/admin/artifact-runs/${runId}/events?tenant_slug=${tenantSlug}`
      : `/admin/artifact-runs/${runId}/events`;
    return httpClient.get<ArtifactRunEventsResponse>(url);
  },

  submitCodingAgentPrompt: async (
    payload: ArtifactCodingPromptRequest,
    tenantSlug?: string,
  ): Promise<ArtifactCodingPromptSubmissionResponse> => {
    const url = tenantSlug
      ? `/admin/artifacts/coding-agent/v1/prompts?tenant_slug=${tenantSlug}`
      : "/admin/artifacts/coding-agent/v1/prompts";
    return httpClient.post<ArtifactCodingPromptSubmissionResponse>(url, payload);
  },

  listCodingAgentChatSessions: async (
    options: { artifactId?: string | null; draftKey?: string | null; scopeMode?: ArtifactCodingScopeMode | null; limit?: number },
    tenantSlug?: string,
  ): Promise<ArtifactCodingChatSession[]> => {
    const params = new URLSearchParams();
    if (options.artifactId) params.set("artifact_id", options.artifactId);
    if (options.draftKey) params.set("draft_key", options.draftKey);
    if (options.scopeMode) params.set("scope_mode", options.scopeMode);
    params.set("limit", String(Math.max(1, Number(options.limit || 25))));
    if (tenantSlug) params.set("tenant_slug", tenantSlug);
    return httpClient.get<ArtifactCodingChatSession[]>(
      `/admin/artifacts/coding-agent/v1/sessions?${params.toString()}`,
    );
  },

  getCodingAgentChatSession: async (
    sessionId: string,
    options: { limit?: number; before_message_id?: string | null; tenantSlug?: string } = {},
  ): Promise<ArtifactCodingChatSessionDetail> => {
    const params = new URLSearchParams();
    params.set("limit", String(Math.max(1, Number(options.limit || 10))));
    if (options.before_message_id) {
      params.set("before_message_id", options.before_message_id);
    }
    if (options.tenantSlug) {
      params.set("tenant_slug", options.tenantSlug);
    }
    return httpClient.get<ArtifactCodingChatSessionDetail>(
      `/admin/artifacts/coding-agent/v1/sessions/${sessionId}?${params.toString()}`,
    );
  },

  getCodingAgentChatSessionActiveRun: async (
    sessionId: string,
    tenantSlug?: string,
  ): Promise<ArtifactCodingActiveRunState> => {
    const url = tenantSlug
      ? `/admin/artifacts/coding-agent/v1/sessions/${sessionId}/active-run?tenant_slug=${tenantSlug}`
      : `/admin/artifacts/coding-agent/v1/sessions/${sessionId}/active-run`;
    return httpClient.get<ArtifactCodingActiveRunState>(url);
  },

  findCodingAgentChatSessionActiveRun: async (
    sessionId: string,
    tenantSlug?: string,
  ): Promise<ArtifactCodingActiveRunState | null> => {
    const url = tenantSlug
      ? `/admin/artifacts/coding-agent/v1/sessions/${sessionId}/active-run?tenant_slug=${tenantSlug}`
      : `/admin/artifacts/coding-agent/v1/sessions/${sessionId}/active-run`;
    const response = await httpClient.requestRaw(url, { method: "GET" });
    if (response.status === 404) {
      return null;
    }
    if (!response.ok) {
      let message = "Request failed";
      try {
        const data = await response.json();
        message = data?.detail || data?.message || message;
      } catch {
        message = response.statusText || message;
      }
      throw new Error(typeof message === "string" ? message : JSON.stringify(message));
    }
    return response.json() as Promise<ArtifactCodingActiveRunState>;
  },

  streamCodingAgentRun: async (runId: string, tenantId?: string | null): Promise<Response> => {
    const streamBase = String(process.env.NEXT_PUBLIC_BACKEND_STREAM_URL || "").trim();
    const backendBase = String(process.env.NEXT_PUBLIC_BACKEND_URL || "").trim();
    const directBackendUrl = /^https?:\/\//i.test(streamBase)
      ? streamBase
      : /^https?:\/\//i.test(backendBase)
        ? backendBase
        : "http://127.0.0.1:8026";
    const authState = useAuthStore.getState();
    const token = authState.token;
    const headers: Record<string, string> = {
      Accept: "text/event-stream",
      "Cache-Control": "no-cache",
    };
    if (token) {
      headers.Authorization = `Bearer ${token}`;
    }
    if (tenantId) {
      headers["X-Tenant-ID"] = tenantId;
    } else if (authState.user?.tenant_id) {
      headers["X-Tenant-ID"] = authState.user.tenant_id;
    }
    const url = new URL(
      `/admin/artifacts/coding-agent/v1/runs/${encodeURIComponent(runId)}/stream`,
      directBackendUrl,
    );
    return fetch(url.toString(), {
      method: "GET",
      headers,
      credentials: "include",
    });
  },

  getCodingAgentRun: async (runId: string, tenantSlug?: string): Promise<ArtifactCodingRun> => {
    const url = tenantSlug
      ? `/admin/artifacts/coding-agent/v1/runs/${runId}?tenant_slug=${tenantSlug}`
      : `/admin/artifacts/coding-agent/v1/runs/${runId}`;
    return httpClient.get<ArtifactCodingRun>(url);
  },

  cancelCodingAgentRun: async (runId: string, tenantSlug?: string): Promise<ArtifactCodingRun> => {
    const url = tenantSlug
      ? `/admin/artifacts/coding-agent/v1/runs/${runId}/cancel?tenant_slug=${tenantSlug}`
      : `/admin/artifacts/coding-agent/v1/runs/${runId}/cancel`;
    return httpClient.post<ArtifactCodingRun>(url, {});
  },

  answerCodingAgentRunQuestion: async (
    runId: string,
    payload: ArtifactCodingAnswerQuestionRequest,
    tenantSlug?: string,
  ): Promise<ArtifactCodingRun> => {
    const url = tenantSlug
      ? `/admin/artifacts/coding-agent/v1/runs/${runId}/answer-question?tenant_slug=${tenantSlug}`
      : `/admin/artifacts/coding-agent/v1/runs/${runId}/answer-question`;
    return httpClient.post<ArtifactCodingRun>(url, payload);
  },

  revertCodingAgentSession: async (
    sessionId: string,
    payload: ArtifactCodingRevertRequest,
    tenantSlug?: string,
  ): Promise<ArtifactCodingChatSessionDetail> => {
    const url = tenantSlug
      ? `/admin/artifacts/coding-agent/v1/sessions/${sessionId}/revert?tenant_slug=${tenantSlug}`
      : `/admin/artifacts/coding-agent/v1/sessions/${sessionId}/revert`;
    return httpClient.post<ArtifactCodingChatSessionDetail>(url, payload);
  },

  listCodingAgentModels: async (): Promise<ArtifactCodingModelOption[]> => {
    return [{ id: null, label: "Auto", description: "Use the default artifact coding model" }];
  },
};
